controllers/users-controllers.js

```diff
@@ -0,0 +1,134 @@
+const { validationResult } = require('express-validator');
+const bcrypt = require('bcryptjs');
+
+const jwt = require('jsonwebtoken');
+
+const User = require('../models/user');
+
+const signup = async (req, res, next) => {
+	const errors = validationResult(req);
+	if (!errors.isEmpty()) {
+		return next(errors);
+	}
+	const { name, email, password, country, phone, newsletters } = req.body;
+
+	let userExists;
+	try {
+		userExists = await User.findOne({ email });
+	} catch (error) {
+		console.log(error);
+		res.status(500).json({
+			msg: 'Could not save user in database'
+		});
+		return next(error);
+	}
+	if (userExists) {
+		res.status(422).json({
+			msg: 'Invalid credentials.'
+		});
+		throw new Error('Invalid credentials.');
+	}
+
+	let hashedPassword;
+	try {
+		hashedPassword = await bcrypt.hash(password, 12);
+	} catch (error) {
+		res.status(500).json({
+			msg: 'Could not save user'
+		});
+		return next(error);
+	}
+
+	const newUser = new User({ name, email, password: hashedPassword, country, phone, newsletters });
+
+	try {
+		await newUser.save();
+	} catch (error) {
+		console.log(error);
+		res.status(500).json({
+			msg: 'Could not save user in database'
+		});
+		return next(error);
+	}
+
+	let token;
+	try {
+		token = jwt.sign({ userId: newUser.id, email: newUser.email }, process.env.JWT_KEY, {
+			expiresIn: '1h'
+		});
+	} catch (err) {
+		console.log(err);
+		res.status(500).json({
+			msg: 'sign up failed. Please try again later'
+		});
+		return next(error);
+	}
+
+	res.status(201).json({
+		userId: newUser.id,
+		email: newUser.email,
+		token
+	});
+};
+
+const login = async (req, res, next) => {
+	const { email, password } = req.body;
+
+	let existingUser;
```

Comment:
this part of the code that is checking if a user exists is used by login and the signup method. I think we could move this logic to an isolated method so we can reuse the logic.